src/iarc7_motion/iarc_tasks/position_holder.py

```diff
@@ -45,62 +47,91 @@ def get_xy_hold_response(self, current_x, current_y, z_velocity=None):
 
     def _current_velocity_callback(self, odometry):
         with self._lock:
-            if odometry.header.frame_id == 'map' and odometry.child_frame_id == 'level_quad':
+            if (odometry.header.frame_id == 'map' 
+                and odometry.child_frame_id == 'level_quad'):
                 self._odometry = odometry
                 if self._hold_x is None:
                     self._hold_x = odometry.pose.pose.position.x
                 if self._hold_y is None:
                     self._hold_y = odometry.pose.pose.position.y
+                if self._last_vel_x is None:
+                    self._last_vel_x = odometry.twist.twist.linear.x
+                if self._last_vel_y is None:
+                    self._last_vel_y = odometry.twist.twist.linear.y
+                if self._last_speed is None:
+                    self._last_speed = (
+                        math.sqrt(odometry.twist.twist.linear.x**2 
+                                  + odometry.twist.twist.linear.y**2))
             else:
                 rospy.logwarn('Received odometry message with incorrect frame and child frames')
 
-    def _calculate_trapezoidal_acceleration(self, distance, speed):
-        # Check if we can decelerate in time
+    def _calculate_trapezoidal_acceleration(self, x, y, distance):
         try:
-            target_acceleration = (speed**2)/(2*distance)
+            # Compute a dot product to find our current speed towards a target
+            # The z term can be discarded since we aren't concerned with the z velocity
+            speed_towards_target = ((self._last_vel_x * x + self._last_vel_y * y)
+                                    / distance)
         except ZeroDivisionError:
-            # We are at the target point, no need to accelerate anywhere
-            return 0.0
-
-        if target_acceleration < self._max_acceleration:
-            if speed < self._max_speed:
-                target_acceleration = self._max_acceleration
+            # Last speed was zero therefore the last speed to the target was 0
+            speed_towards_target = 0
+
+        if speed_towards_target > 0:
+            # Check if we can decelerate in time
+            try:
+                target_acceleration = (speed_towards_target**2)/(2*distance)
+            except ZeroDivisionError:
+                # We are at the target point, no need to accelerate anywhere
+                return 0.0
+
+            if target_acceleration < self._max_acceleration:
```

Comment:
That is not a limit check. If the target acceleration calculated above is more than the max acceleration it means that we've crossed the point where we can decelerate in time to stop at the target position. I think the code is confusing as written so I'll adjust it.